postgres_composite_types/__init__.py

```diff
@@ -225,6 +210,58 @@ def __new__(mcs, name, bases, attrs):
 
         return new_cls
 
+    def __init__(cls, name, bases, attrs):
+        super().__init__(name, bases, attrs)
+        if name == 'CompositeType':
+            return
+
+        # Register the type on the first database connection
+        connection_created.connect(receiver=cls.database_connected,
+                                   dispatch_uid=cls._meta.db_type)
+
+    def database_connected(cls, signal, sender, connection, **kwargs):
+        if isinstance(connection, PostgresDatabaseWrapper):
+            # Try to register the type. If the type has not been created in a
+            # migration, the registration should be retried later.
+            cls.register_composite(connection, retry=True)
+
+        # Disconnect the signal now - only need to register types on the
+        # initial connection
+        connection_created.disconnect(cls.database_connected,
+                                      dispatch_uid=cls._meta.db_type)
+
+    def register_composite(cls, connection, retry=False):
+        """
+        Register this CompositeType with Postgres.
+
+        If the CompositeType does not yet exist in the database, this will
+        fail.  Hopefully a migration will come along shortly and create the
+        type in the database. If `retry` is True, this CompositeType will try
+        to register itself again after the type is created.
+        """
+
+        try:
+            with connection.temporary_connection() as cur:
+                register_composite(cls._meta.db_type, cur, globally=True)
+        except ProgrammingError:
+            if retry:
+                composite_type_created.connect(post_migrate_register, sender=cls)
+            else:
+                raise
+
+
+def post_migrate_register(signal, sender, connection, **kwargs):
```

Comment:
Signals are run synchronously, so you could already run a further data migration just after this one. Directly calling `self.Meta.model.register_composite(connection)` would be sensible and simpler though. I will do that.